akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala

```diff
@@ -0,0 +1,733 @@
+
+  private[projection] implicit val executionContext: ExecutionContext = system.executionContext
+
+  // The OffsetStore instance is used by a single projectionId and there shouldn't be many concurrent
+  // calls to methods that access the `state`, but for example validate (load) may be concurrent
+  // with save. Therefore, this state can be updated concurrently with CAS retries.
+  private val state = new AtomicReference(State.empty)
+
+  // Transient state of inflight pid -> seqNr (before they have been stored and included in `state`), which is
+  // needed for at-least-once or other projections where the offset is saved afterwards. Not needed for exactly-once.
+  // This can be updated concurrently with CAS retries.
+  private val inflight = new AtomicReference(Map.empty[Pid, SeqNr])
+
+  private def timestampOf(persistenceId: String, sequenceNr: Long): Future[Option[Instant]] = {
+    sourceProvider match {
+      case Some(timestampQuery: EventTimestampQuery) =>
+        timestampQuery.timestampOf(persistenceId, sequenceNr)
+      case Some(timestampQuery: akka.persistence.query.typed.javadsl.EventTimestampQuery) =>
+        import scala.jdk.FutureConverters._
+        import scala.jdk.OptionConverters._
+        timestampQuery.timestampOf(persistenceId, sequenceNr).asScala.map(_.toScala)
+      case Some(_) =>
+        throw new IllegalArgumentException(
+          s"Expected BySlicesSourceProvider to implement EventTimestampQuery when TimestampOffset is used.")
+      case None =>
+        throw new IllegalArgumentException(
+          s"Expected BySlicesSourceProvider to be defined when TimestampOffset is used.")
+    }
+  }
+
+  def getState(): State =
+    state.get()
+
+  def getInflight(): Map[Pid, SeqNr] =
+    inflight.get()
+
+  // This is used by projection management and returns latest offset
+  def getOffset[Offset](): Future[Option[Offset]] = {
+    // FIXME in r2dbc this will reload via readOffset if no state
+    Future.successful(Some(getState().latestOffset).map(_.asInstanceOf[Offset]))
+  }
+
+  def readOffset[Offset](): Future[Option[Offset]] = {
+    // look for TimestampOffset first since that is used by akka-persistence-dynamodb,
+    // and then fall back to the other more primitive offset types
+    sourceProvider match {
+      case Some(_) =>
+        readTimestampOffset().map { offsetBySlice =>
+          if (offsetBySlice.offsets.isEmpty) None
+          else Some(offsetBySlice.asInstanceOf[Offset])
+        }(ExecutionContext.parasitic)
+      case None =>
+        // FIXME primitive offsets not supported, maybe we can change the sourceProvider parameter
+        throw new IllegalStateException("BySlicesSourceProvider is required. Primitive offsets not supported.")
+    }
+  }
+
+  private def readTimestampOffset(): Future[TimestampOffsetBySlice] = {
+    val oldState = state.get()
+    // retrieve latest timestamp for each slice, and use the earliest
+    val futTimestamps =
```

Comment:
This is conceivably 1024 queries fired off right at startup, wonder if it's better to do this in a stream with mapAsync and limited parallelism?

Reply:
Yes, needing the separate queries is a weak point for the plugin. The events-by-slices queries are also all separate and in parallel. How many queries that actually run in parallel should come down to the concurrency settings for the DynamoDB client. Not sure if @patriknw had any other thoughts at the time?

(And let's do any improvements in separate PRs. With this one as the initial copy over.)

Reply:
Is the concern that it creates a burst that is not desired? I don't know if that is a problem. Create an issue/pr if that has been seen as a real problem when testing.

events-by-slices queries are more separate and therefore more spread over time

Reply:
I'll have a PR up for at least one of these when this is merged

Reply:
sounds good to add the bounded parallelism, thanks